src/nupic/algorithms/Connections.hpp

```diff
@@ -41,11 +41,11 @@ namespace nupic {
 namespace algorithms {
 namespace connections {
 
-typedef UInt32 CellIdx;
-typedef UInt16 SegmentIdx; /** Index of segment in cell. */
-typedef UInt16 SynapseIdx; /** Index of synapse in segment. */
-typedef UInt32 Segment;    /** Index of segment's data. */
-typedef UInt32 Synapse;    /** Index of synapse's data. */
+typedef UInt32 CellIdx;  //TODO instead of typedefs, use templates for proper type-checking
+typedef unsigned char SegmentIdx; /** Index of segment in cell. */
+typedef unsigned char SynapseIdx; /** Index of synapse in segment. */
+typedef UInt16 Segment;    /** Index of segment's data. */
+typedef UInt16 Synapse;    /** Index of synapse's data. */
```

Comment:
note, a WIP commit changes these to much lower values. 
I'll need to asses the impact, suprisingly, the lower values (compared to UInt) do not gain a significant performance gain (possibly some forgotten type-casting happens?) 

Unit test SP:: ExactOutput "breaks" as stays stuck, does not finish, nor fail. I don;t know why

Reply:
The size of this data type controls the maximum number of synapses & segments.
So with an 8-bit integer, it can only have a 255 synapses, maximum.
16-bits as shown here allows a maximum of 65,536 synapses.

I will give this a more thorough review within a day or two.

Reply:
> So with an 8-bit integer, it can only have a 256 synapses, maximum.
16-bits as shown here allows a maximum of 65,536 synapses.

yes. Are you suggesting the 16b should match the 8-bit? As SynapseIdx is "synapses per segment", and `Synapse` is a flat array/index to all existing synapses (their data). But there can be a number of segments in cells, so Synapses >> SynapseIdx.

Reply:
> Are you suggesting

No i just wanted to clarify.
What u have here looks good, as long as it doesnt overflow.

Reply:
And if it does overflow then we can easily fix it by changing these typedefs which after this PR should be used much more consistently. :)